x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java

```diff
@@ -152,97 +181,56 @@ public void roles(Set<String> roleNames, FieldPermissionsCache fieldPermissionsC
-                    negativeLookupCache.addAll(missing);
-                    return builtInRoleDescriptors;
-                }).run();
-            } else {
-                logger.debug(() ->
-                        new ParameterizedMessage("Requested roles [{}] do not exist", Strings.collectionToCommaDelimitedString(missing)));
-                negativeLookupCache.addAll(missing);
-                roleDescriptorActionListener.onResponse(Collections.unmodifiableSet(builtInRoleDescriptors));
-            }
-        } else {
-            roleDescriptorActionListener.onResponse(Collections.unmodifiableSet(builtInRoleDescriptors));
-        }
+        loadRoleDescriptorsAsync(filteredRoleNames, rolesResultListener);
     }
 
-    private Set<RoleDescriptor> getBuiltInRoleDescriptors(Set<String> roleNames) {
-        final Set<RoleDescriptor> descriptors = reservedRolesStore.roleDescriptors().stream()
-                .filter((rd) -> roleNames.contains(rd.getName()))
-                .collect(Collectors.toCollection(HashSet::new));
-        if (descriptors.size() > 0) {
-            logger.debug(() -> new ParameterizedMessage("Roles [{}] are builtin roles", names(descriptors)));
-        }
-        final Set<String> difference = difference(roleNames, descriptors);
-        if (difference.isEmpty() == false) {
-            final Set<RoleDescriptor> fileRoles = fileRolesStore.roleDescriptors(difference);
-            logger.debug(() ->
-                    new ParameterizedMessage("Roles [{}] were resolved from [{}]", names(fileRoles), fileRolesStore.getFile()));
-            descriptors.addAll(fileRoles);
-        }
-
-        return descriptors;
+    private void loadRoleDescriptorsAsync(Set<String> roleNames, ActionListener<RolesRetrievalResult> listener) {
+        final RolesRetrievalResult rolesResult = new RolesRetrievalResult();
+        final List<BiConsumer<Set<String>, ActionListener<RoleRetrievalResult>>> asyncRoleProviders =
+            licenseState.isCustomRoleProvidersAllowed() ? allRoleProviders : builtInRoleProviders;
+
+        final ActionListener<RoleRetrievalResult> descriptorsListener =
+            ContextPreservingActionListener.wrapPreservingContext(ActionListener.wrap(ignore -> {
+                    rolesResult.setMissingRoles(roleNames);
+                    listener.onResponse(rolesResult);
+                }, listener::onFailure), threadContext);
+
+        final Predicate<RoleRetrievalResult> iterationPredicate = result -> roleNames.isEmpty() == false;
+        new IteratingActionListener<>(descriptorsListener, (rolesProvider, providerListener) -> {
+            // try to resolve descriptors with role provider
+            rolesProvider.accept(roleNames, ActionListener.wrap(result -> {
+                if (result.isSuccess()) {
+                    logger.debug(() -> new ParameterizedMessage("Roles [{}] were resolved by [{}]",
+                        names(result.getDescriptors()), rolesProvider));
```

Comment:
I don’t think we should do this. It’s a micro optimization and since it is at debug level we would either complicate the code to build a list conditionally or unnecessarily build the list of names all the time. We use a supplier so that we only build the names when we actually log this statement.